server/src/main/java/org/elasticsearch/search/slice/SliceBuilder.java

```diff
@@ -273,6 +315,23 @@ public Query toFilter(QueryShardContext context, int shardId, int numShards) {
         return new MatchAllDocsQuery();
     }
 
+    /**
+     * Returns the {@link GroupShardsIterator} for the provided <code>request</code>.
+     */
+    private GroupShardsIterator<ShardIterator> buildShardIterator(ClusterService clusterService, ShardSearchRequest request) {
+        final ClusterState state = clusterService.state();
+        final Settings settings = clusterService.getSettings();
+        final String[] indices;
+        if (request instanceof IndicesRequest) {
+            indices = ((IndicesRequest) request).indices();
```

Comment:
Maybe we can change the `ShardSearchRequest` and instead of sending the global routing for the request we send the list of extracted routings for the requested index. `ShardSearchRequest#routings` and when there is no alias routing it returns the global routing in a singleton ?

Reply:
yeah I think the routing needs to be resolved on the coordinator